src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/ApprovedPremisesOfflineApplicationsSeedJob.kt

```diff
@@ -0,0 +1,55 @@
+package uk.gov.justice.digital.hmpps.approvedpremisesapi.seed
+
+import org.slf4j.LoggerFactory
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.ServiceName
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.OfflineApplicationEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.OfflineApplicationRepository
+import java.time.OffsetDateTime
+import java.util.UUID
+
+class ApprovedPremisesOfflineApplicationsSeedJob(
+  fileName: String,
+  private val offlineApplicationRepository: OfflineApplicationRepository,
+) : SeedJob<OfflineApplicationsSeedCsvRow>(
+  fileName = fileName,
+  requiredColumns = 12,
+) {
+  private val log = LoggerFactory.getLogger(this::class.java)
+
+  override fun verifyPresenceOfRequiredHeaders(headers: Set<String>) {
+    val missingHeaders = requiredHeaders() - headers
+
+    if (missingHeaders.any()) {
+      throw RuntimeException("required headers: $missingHeaders")
+    }
+  }
+
+  override fun deserializeRow(columns: Map<String, String>) = OfflineApplicationsSeedCsvRow(
+    crn = columns["crn"]!!,
+  )
+
+  override fun processRow(row: OfflineApplicationsSeedCsvRow) {
+    val existingOfflineApplications = offlineApplicationRepository.findAllByServiceAndCrn(ServiceName.approvedPremises.value, row.crn)
+
+    if (existingOfflineApplications.none()) {
+      offlineApplicationRepository.save(
```

Comment:
This is going to be potentially quite slow with a lot of CRNs, could we get all the CRNs with existing bookings with an `IN` query, diff the array of CRNs, create an array of `OfflineApplicationEntity`s and then save all in one?

Some psuedocode:

```
val crns = content_of_csv.map { it.crn }
val existingBookings = offlineApplicationRepository.findAllWhereCrnIn(crns)
val crnsWithExistingBookings = existingBookings.map { it.crn }
val crnsToCreateBookingsFor = crns - crnsWithExistingBookings
val applications = crnsToCreateBookingsFor.map {
  OfflineApplicationEntity(
    id = UUID.randomUUID(),
    crn = it,
    service = ServiceName.approvedPremises.value,
    createdAt = OffsetDateTime.now(),
  )
}
offlineApplicationRepository.save(applications)
```

Reply:
Ah, course. Nah, I think this approach is fine in that case